Common/ExtendedDictionary.cs

```diff
@@ -0,0 +1,341 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using Python.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Interfaces;
+using System.Collections;
+
+namespace QuantConnect
+{
+    /// <summary>
+    /// Provides a base class for types holding instances keyed by <see cref="Symbol"/>
+    /// </summary>
+    public abstract class ExtendedDictionary<T> : IExtendedDictionary<Symbol, T>
+    {
+        /// <summary>
+        /// Removes all items from the <see cref="T:System.Collections.Generic.ICollection`1"/>.
+        /// </summary>
+        /// <exception cref="T:System.NotSupportedException">The <see cref="T:System.Collections.Generic.ICollection`1"/> is read-only. </exception>
+        public virtual void Clear()
+        {
+            if (IsReadOnly)
+            {
+                throw new InvalidOperationException($"Clear/clear method call is an invalid operation. {GetType().Name} is a read-only collection.");
+            }
+            throw new NotImplementedException("Types deriving from 'BaseDictionary' must implement the 'void Clear() method.");
```

Comment:
nit - There are a few exception where we should replace `'BaseDictionary'` for `'ExtendedDictionary'` after we renamed